util: Don't reference errno when pthread fails.

miztake opened the pull request:
Pthread library does not set errno.
Pthread library's errno is returned by return value.

practicalswift commented:
Concept ACK: nice catch and nice first-time contribution!

May I ask how did you found this issue? :)

Warm welcome as a contributor @miztake :)

pstratem commented:
ACK c7601aba89972f71464c9809cc3607b96b5975ab

you can move the declaration into the if statement like this:
`if ((int rc = pthread_setschedparam(pthread_self(), SCHED_BATCH, &param)) != 0) {`

hebasto commented on src/util/system.cpp:
```diff
@@ -1163,8 +1163,9 @@ void ScheduleBatchPriority()
     const static sched_param param{};
-    if (pthread_setschedparam(pthread_self(), SCHED_BATCH, &param) != 0) {
-        LogPrintf("Failed to pthread_setschedparam: %s\n", strerror(errno));
+    int rc;
+    if ((rc = pthread_setschedparam(pthread_self(), SCHED_BATCH, &param)) != 0) {
```
nit: could look clearer
```suggestion
    const int rc = pthread_setschedparam(pthread_self(), SCHED_BATCH, &param);
    if (rc != 0) {
```

...  until switching to C++17 :)

miztake replied:
Hi, @hebasto

Thank you for suggesting change.
I accept your suggested change.

hebasto reviewed (commented):
ACK c7601aba89972f71464c9809cc3607b96b5975ab, tested on Linux Mint 19.3 (x86_64).

From the [`pthread_setschedparam()` docs](https://man7.org/linux/man-pages/man3/pthread_setschedparam.3.html):
> On success, these functions return 0; on error, they return a nonzero error number.

hebasto commented:
> you can move the declaration into the if statement like this:
> `if ((int rc = pthread_setschedparam(pthread_self(), SCHED_BATCH, &param)) != 0) {`

It won't compile :) (at least on GCC 7.5).

miztake commented:
Hi, @practicalswift

> May I ask how did you found this issue? :)

Thank you for your comment.
I didn't actually output this log.
It's a common bug when using pthread and I found it in a code review.
( just greped "pthread" :) )

donaloconnor commented:
ACK 0e7b64b64db00d03489a72c988dc7ef53433f69d

hebasto reviewed (approved):
ACK 0e7b64b64db00d03489a72c988dc7ef53433f69d.

Mind squashing commits into one commit?

kristapsk commented:
Concept ACK, checked `man pthread_setschedparam`, change seems correct, but commits should be squashed into one.

miztake committed:
util: Don't reference errno when pthread fails.

Pthread library does not set errno.
Pthread library's errno is returned by return value.

Co-authored-by: Hennadii Stepanov <32963518+hebasto@users.noreply.github.com>

miztake commented:
Hi, @kristapsk

Thank you for your comment.
I squashed the commits into one.

hebasto reviewed (approved):
ACK cb38b069b0f41b1a26264784b1c1303c8ac6ab08, only squashed commits since the [previous](https://github.com/bitcoin/bitcoin/pull/19194#pullrequestreview-425831739) review.

MarcoFalke commented:
review ACK cb38b069b0f41b1a26264784b1c1303c8ac6ab08

practicalswift commented:
ACK cb38b069b0f41b1a26264784b1c1303c8ac6ab08 -- patch looks correct

laanwj commented:
Poshumous ACK, good catch!
